adapters/coap-vertx-base/src/main/java/org/eclipse/hono/adapter/coap/AbstractVertxBasedCoapAdapter.java

```diff
@@ -1112,4 +1096,95 @@ private TtdStatus getTtdStatus(final CoapContext context) {
+                ? Future.succeededFuture(cmdResponseOrNull)
+                : Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
+                        String.format("command-request-id [%s] or status code [%s] is missing/invalid",
+                                commandRequestId, responseStatus)));
+
+        final int payloadSize = Optional.ofNullable(payload).map(Buffer::length).orElse(0);
+        return CompositeFuture.all(tenantTracker, commandResponseTracker)
+                .compose(commandResponse -> {
+                    final Future<JsonObject> deviceRegistrationTracker = getRegistrationAssertion(
+                            device.getTenantId(),
+                            device.getDeviceId(),
+                            authenticatedDevice,
+                            currentSpan.context());
+                    final Future<Void> tenantValidationTracker = CompositeFuture
+                            .all(isAdapterEnabled(tenantTracker.result()),
+                                    checkMessageLimit(tenantTracker.result(), payloadSize, currentSpan.context()))
+                            .map(ok -> null);
+
+                    return CompositeFuture.all(tenantValidationTracker, deviceRegistrationTracker)
+                            .compose(ok -> sendCommandResponse(device.getTenantId(), commandResponseTracker.result(), currentSpan.context()))
+                            .map(delivery -> {
+                                log.trace("delivered command response [command-request-id: {}] to application",
+                                        commandRequestId);
+                                currentSpan.log("delivered command response to application");
+                                currentSpan.finish();
+                                metrics.reportCommand(
+                                        Direction.RESPONSE,
+                                        device.getTenantId(),
+                                        tenantTracker.result(),
+                                        ProcessingOutcome.FORWARDED,
+                                        payloadSize,
+                                        context.getTimer());
+                                context.respondWithCode(ResponseCode.CHANGED);
+                                return ResponseCode.CHANGED;
+                            });
+                }).otherwise(t -> {
+                    log.debug("could not send command response [command-request-id: {}] to application",
+                            commandRequestId, t);
+                    TracingHelper.logError(currentSpan, t);
+                    currentSpan.finish();
+                    metrics.reportCommand(
+                            Direction.RESPONSE,
+                            device.getTenantId(),
+                            tenantTracker.result(),
+                            ProcessingOutcome.from(t),
+                            payload.length(),
```

Comment:
payloadSize?

Reply:
Here it seems to be cosmetic, payload will be always provided but may be `null`.
Anyway, if will remove the `payloadSize` and use always the `payload.length()`, though it's very straight forward, when payload is never `null`.
Just to mention: the "payload.length()" is copied form http, where payload may be `null`. so it may cause troubles there.